alpenhorn/io/nearline.py

```diff
@@ -185,6 +266,92 @@ def check(self, copy: ArchiveFileCopy) -> None:
         # Otherwise, use the DefaultIO method to do the check
         super().check(copy)
 
+    def auto_verify(self, copy: ArchiveFileCopy) -> None:
+        """Auto-verify a file in nearline.
+
+        If the file happens to be restored, we simply run check() on
+        it.  If it's released, instead we do the following:
+
+        - trigger a restore the file
+        - wait for the file to be restored
+        - run the check()
+        - release the file again
+
+        The last step here is to keep auto-verifcation from replacing
+        all the newly added data from nearline (which are presumably
+        more interesting that some random files being auto-verified).
+
+        Parameters
+        ----------
+        copy : ArchiveFileCopy
+            the file copy to auto-verify
+        """
+        # The trivial case.
+        if not pathlib.Path(copy.path).exists():
+            # Only update if this is surprising (which it probably
+            # is.)
+            if copy.has_file != "N":
+                log.warning(
+                    "File copy missing during auto-verify: "
+                    f"{copy.file.path}.  Updating database."
+                )
+                ArchiveFileCopy.update(has_file="N").where(
+                    ArchiveFileCopy.id == copy.id
+                ).execute()
+            return
+
+        # Are we restored?  Note: we're interested in the actual
+        # state of the file, not the state as recorded in the database.
+        #
+        # If we are, we can just use the DefaultIO's check() method
+        # and not worry about any of the stuff going on below.
+        if not self._lfs.hsm_released(copy.path):
+            return super().check(copy)
+
+        def _async(
```

Comment:
What would happen if this is called on the same file multiple times? Since `QueryWalker` can return duplicates and this is asynchronous, it's entirely possible that we could call `auto_verify` on the same file twice

Reply:
Good point.

I think verification of the same file twice woudl work out fine.  It'd be a waste of effory (MD5-ing twice) but the last task running would end up winning.  Even if they overwrite each other, they should come to the same conclusion, so I think nothing too bad will happen.

Still, might be worth discarding duplicates from the list returned by the query walker.

However, that does make me think: what about the case when you turn on auto-verify on.a node with only one file?  Even if we do discard duplicates in a single update loop, alpenhorn's still just going to grind md5-ing that file over and over again for lack of anything else to do.  It might be worth adding a minimum wait time before re-verifying (i.e. skip everything with last_update less than a day (or week or something) old.

Reply:
Agreed, something like a day or two would probably be fine

Reply:
Both done:
* duplicates are skipped
* "new" files are not rechecked, where newness can be specified by the "auto_verify_min_days" config option and defaults to 7 days.  Because verification updates last_update, this will prevent re-verification from happening too often.